tests/shared/src/test/scala/cats/tests/HashMapSuite.scala

```diff
@@ -0,0 +1,292 @@
+      val hashMap = HashMap.fromSeq(kvs)
+      assert(hashMap.isEmpty === kvs.isEmpty)
+      assert(hashMap.nonEmpty === kvs.nonEmpty)
+    }
+  }
+
+  test("===") {
+    forAll { (kvs: List[(Int, String)]) =>
+      val left = HashMap.fromSeq(kvs)
+      val right = HashMap.fromSeq(kvs)
+      assert(left === right)
+    }
+
+    forAll { (leftKvs: List[(Int, String)], rightKvs: List[(Int, String)]) =>
+      val left = HashMap.fromSeq(leftKvs)
+      val right = HashMap.fromSeq(rightKvs)
+      assert((distinctBindings(leftKvs) === distinctBindings(rightKvs)) === (left === right))
+    }
+  }
+
+  test("size") {
+    assert(HashMap.empty[Int, String].size === 0)
+    assert(HashMap(1 -> "a", 2 -> "b", 3 -> "c").size === 3)
+    assert(HashMap("Aa" -> 1, "BB" -> 2).size == 2)
+
+    forAll { (kvs: List[(Int, String)]) =>
+      val hashMap = HashMap.fromSeq(kvs)
+      assert(hashMap.size === distinctBindings(kvs).size)
+    }
+  }
+
+  test("get") {
+    forAll { (kvs: List[(Int, String)]) =>
+      val hashMap = HashMap.fromSeq(kvs)
+      distinctBindings(kvs).foreach { case (k, v) =>
+        hashMap.get(k) === Some(v)
+      }
+    }
+  }
+
+  test("concat") {
+    forAll { (leftKvs: List[(Int, String)], rightKvs: List[(Int, String)]) =>
+      val distinctKvs = distinctBindings(leftKvs ++ rightKvs)
+      val left = HashMap.fromSeq(leftKvs)
+      val right = HashMap.fromSeq(rightKvs)
+      val both = left.concat(right)
+      assert(both === HashMap.fromSeq(leftKvs ++ rightKvs))
+      distinctKvs.foreach { case (k, v) =>
+        assert(both.contains(k))
+        assert(both.get(k) === Some(v))
+      }
+    }
+  }
+
+  property("Empty HashMap never contains") {
+    forAll { (i: Int) =>
+      assert(!HashMap.empty[Int, String].contains(i))
+    }
+  }
+
+  property("Empty HashMap add consistent with contains") {
+    forAll { (i: Int, s: String) =>
+      assert(HashMap.empty[Int, String].add(i, s).contains(i))
+    }
+  }
+
+  property("Empty HashMap add and remove consistent with contains") {
+    forAll { (i: Int, s: String) =>
+      assert(!HashMap.empty[Int, String].add(i, s).remove(i).contains(i))
+    }
+  }
+
+  property("fromSeq consistent with contains") {
+    forAll { (kvs: List[(Int, String)]) =>
+      val hashMap = HashMap.fromSeq(kvs)
+
+      kvs.foreach { case (k, _) =>
+        assert(hashMap.contains(k))
+      }
+    }
+  }
+
+  property("remove consistent with contains") {
+    forAll { (kvs: List[(Int, String)]) =>
+      val distinctKvs = distinctBindings(kvs)
+      val hashMap = HashMap.fromSeq(kvs)
+
+      distinctKvs.foldLeft(hashMap) { case (hm, (i, _)) =>
+        if (!hm.contains(i)) {
+          println(hm)
+          println(i)
+          println(hm.contains(i))
+        }
+        assert(hm.contains(i))
+        assert(!hm.remove(i).contains(i))
+        hm.remove(i)
+      }
+
+      ()
+    }
+  }
+
+  property("add and remove with collisions consistent with contains") {
```

Comment:
you can actually nest forAll, so we could do this:

```scala
def genKey[K](map: HashMap[K, Any])(implicit k: Gen[K]): Gen[K] =
  if (map.isEmpty) k
  else Gen.oneOf(
    Gen.oneOf(map.keys.toSeq),
    k
  )

forAll { (h: HashMap[Int, String]) =>
  forAll(genKey(h)) { (key: Int) =>
    // now key is a collision half the time so we really exercise both the collision and non collision cases.
  }
}
```

Reply:
I did something slightly different in the end by using the test structure that you describe, but permuting an existing key to produce a colliding string instead of an exact match. Let me know what you think!